### ISSUE TITLE
mpv and retina display

### ISSUE THREAD

**Issue by sosia** [mpv-player/mpv#3271]
Hello everyone,
I'm using the latest master of mpv on the latest release of El Capitan on a 2015 retina MBP: being the default resolution of 1440x900 a little too small for my taste, I've switched the apparent resolution to the middle "scaled - more space" one (1680x1050) via System Preference.

I was of course expecting it to have an impact on resources, but it was way more than I thought using mpv: e.g., a 60fps clip is smooth with the iGPU with the default resolution, but it drops a lot of frame with the other slightly bigger one. 

I tried to play some video to investigate a bit, and the logs's show how in some of the cases the output image is even bigger than the actual max resolution of the screen, 2880 x 1800, and in the last one considerably so. Again, even if this is not entirely surprising given the retina doubling, talking with @haasn he said it should be treated as a bug, if I understood his words correctly.

> Windowed 720p source:

```
[vo/opengl-hq] Resize: 2560x1440
[vo/opengl-hq] Window size: 2560x1440
[vo/opengl-hq] Video source: 1280x720 (1:1)
[vo/opengl-hq] Video display: (0, 0) 1280x720 -> (0, 0) 2560x1440
```

> Fullscreen 720p source:

```
[vo/opengl-hq] Resize: 3360x2100
[vo/opengl-hq] Window size: 3360x2100
[vo/opengl-hq] Video source: 1280x720 (1:1)
[vo/opengl-hq] Video display: (0, 0) 1280x720 -> (0, 105) 3360x1890
```

> Windowed 1080p source:

```
[vo/opengl-hq] Resize: 3840x2160
[vo/opengl-hq] Window size: 3840x2160
[vo/opengl-hq] Video source: 1920x1080 (1:1)
[vo/opengl-hq] Video display: (0, 0) 1920x1080 -> (0, 0) 3840x2160
```

Nonetheless, overall this has a huge impact on resources and theoretically on video quality (albeit I somewhat doubt that's actually noticeable, in fact the OS is resizing the image independently from the mpv's scalers).
Moreover, subsequently, the `screenshot window` command outputs a resized image instead of one with the actual resolution of the source.

Before reporting this issue I researched on my own and I saw that this was discussed a couple other times on the bug tracker or in IRC, and I understand that the current mpv's behaviour is perfectly consistent with Apple's guidelines (VLC does the same, AFAIK). Yet, for a video player like mpv, which options could already be taxing on the GPU, and well liked for its high quality output, this could lead to inconvenient outcomes.

In conclusion, if there is actually a way, it would be very good if mpv could draw its window independently from the OS' retina scaling.

If you need further logs, information or testing, I'll try to provide them all quickly.

**Comment by haasn** [comment]
To perhaps clarify my stance somewhat:

The concrete bug here is that mpv is scaling the video up past the actual final resolution, and the OS X compositor is scaling the result of the mpv window back down.

The intended result is that mpv directly scales to the final resolution, with mpv's output pixels displayed as-is (1:1 with no further transformations) on the actual display.

This is not only important for resource usage but also necessary for mpv's scaling algorithms to work correctly, as well as for dithering and color accuracy.

**Comment by ChrisK2** [comment]
> if mpv could draw its window independently from the OS' retina scaling.

I'm not aware of this being possible without mpv entering an exclusive full-screen mode and/or switching the display resolution back to "virtual" 1440x900

**Comment by Argon-** [comment]
> I'm not aware of this being possible without mpv entering an exclusive full-screen mode and/or switching the display resolution back to "virtual" 1440x900

Could this be done as an option?

**Comment by Akemi** [comment]
i don't think an exclusive fullscreen will help here since it doesn't change the underlying virtual desktop resolution. so the only real fix would be changing the desktop resolution in fullscreen to a standard hidpi one (2x scaling) or a standard native resolution. i am not sure if such a functionality should be part of mpv, but i can totally see the reasoning. further problems probably will occur since apple decided to change the standard scaling from "Best" (2x scaling) to ["More Room" on the new MBP](https://9to5mac.com/2016/12/02/15-inch-macbook-pro-screen-resolution-blurry/).

i know @pigoz had a [script](https://github.com/pigoz/dotfiles/blob/master/mpv/scripts/refresh.lua) ([tool](https://github.com/jhford/screenresolution)) to change screen resolution based on fps. would an external script be a viable solution for this or should we really integrate such a functionality directly into mpv?

**Comment by Argon-** [comment]
Well, personally, I'd certainly appreciate if my video player could do this on its own.
Regarding the linked script using the tool `screenresolution`: this doesn't support HiDPI resolutions.

**Comment by Hrxn** [comment]
> [..] since apple decided to change the standard scaling from "Best" (2x scaling) to "More Room" on the new MBP.

FWIW, the new default was already changed with the Macbook refresh (2015). Based on their telemetry data, apparently a majority of users prefers it that way.
